text/0121-data-types.md

```diff
@@ -0,0 +1,435 @@
+
+`"Bruce Willis,Vin Diesel"`
+
+In the example above, the distance between `Bruce` and `Willis` is equal to `1`. The distance between `Vin` and `Diesel` is equal to `1` too.
+But, the distance between `Bruce` and `Vin` is equal to `8`. The same goes for `Bruce` and `Diesel`, or `Willis` and `Vin`, or also `Willis` and `Diesel`.
+
+Let's see another example. Given two documents:
+
+```json
+[
+    {
+        "movie_id": "001",
+        "description": "Bruce.Willis"
+    },
+    {
+        "movie_id": "002",
+        "description": "Bruce super Willis"
+    }
+]
+```
+
+When making a query on `Bruce Willis`, `002` will be the first document returned and `001` will be the second one.
+This will happen because the proximity distance between `Bruce` and `Willis` is equal to `2` in the document `002` whereas the distance between `Bruce` and `Willis` is equal to `8` in the document `001` since the full stop is a hard space.
+
+### 3.2. Array Tokenization
+
+An array is recursively broken into separate string tokens, which means separate words. After the tokenizing process, each word is indexed and stored in the global dictionary of the corresponding index.
+
+#### 3.2.1. Examples
+
+The following input:
+
+```json
+[
+    [
+        "Bruce Willis",
+        "Vin Diesel"
+    ],
+    "Kung Fu Panda"
+]
+```
+
+Will be processed as if all elements were arranged at the same level:
+
+```json
+"Bruce Willis. Vin Diesel. Kung Fu Panda."
+```
+
+The strings above will be separated by soft and hard spaces exactly as explained in the string example.
+
+> See [3.1.1. Examples section](#311-examples).
+
+### 3.3. Object
+
+#### 3.3.1. Example
+
+The following input:
+
+```json
+{
+    "movie_id": "1564saqw12ss",
+    "title": "Kung fu Panda"
+}
+```
+
+In the example above, `movie_id`, `1564saqw12ss`, `title`, `Kung fu Panda` are all considered sentences. The colon `:` and comma `,` characters are used as separators.
+
+`"movie_id. 1564saqw12ss. title. Kung fu Panda."`
+
+These sentences will be separated by soft and hard spaces precisely as explained in the string example.
+
+### 3.4. Nested Structures
+
+Nested structures (e.g. `Object`, `Array of Objects`, etc) are internally flattened to a document's root level.
+
+It allows expressing a nested field in all Meilisearch parameters that accept document attributes.
+
+Meilisearch accepts the `.` (dot-notation) to express a nested field location in a document structure.
+
+### 3.4.1. Examples
+
+#### 3.4.1.1. Object
+
+The following JSON document:
+
+```json
+    {
+        "a": {
+            "b": "c",
+            "d": "e",
+            "f": "g"
+        }
+    }
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": "c",
+    "a.d": "e",
+    "a.f": "g"
+}
+```
+
+#### 3.4.1.2. Array of objects
+
+The following JSON document:
+
+```json
+{
+    "a": [
+        { "b": "c" },
+        { "b": "d" },
+        { "b": "e" },
+    ]
+}
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": ["c", "d", "e"],
+}
+```
```

Comment:
There is a small difference between your two examples. In the first example, the key are unique per object, the objects are contained within an array.

So we are able to make the value of each searchable object available via `a.b`. 

Imagine that you have an index to find cinemas around you and being to search them by movies title and related genres. He will want to be able to facet and find all the genres rather than the last one. This would be a big limitation and the nested fields would be of little use if Meilisearch only kept the last one in this case.

```json
[
    {
        "id": 0,
        "label": "Pathé Gaumont Paris 13"
        "movies": [
            {
                "title": "Cool Movie 1",
                "genres": ["Summer", "Food", "Buddy"]
            },
            {
                "title": "Wayne's World",
                "genres": ["Stairway To Heaven", "Pepsi"]
            }
        ]
    },
    {
       "id": 1,
       "label": "UGC Paris 13",
       "movies": [
          {
              "title": "Dune",
              "genres": ["Space Opera"]
          }
       ]
    }
]
```

With your first example, the engine will only keep the last movie of the first cinema. It will not work.

While on your second example, it's why the engine does. The fields are located in the same "space" and the last one will be kept.

e.g. Using the last example, if we had a `movie` object containing 2 `genres` attributes

```json
[
    {
        "id": 0,
        "label": "Pathé Gaumont Paris 13"
        "movies": [
            {
                "title": "Cool Movie 1",
                "genres": ["Summer", "Food", "Buddy"]
            },
            {
                "title": "Wayne's World",
                "genres": ["Stairway To Heaven", "Pepsi"]
            }
        ]
    },
    {
       "id": 1,
       "label": "UGC Paris 13",
       "movies": [
          {
              "title": "Dune",
              "genres": ["Space Opera"],
              "genres": ["Sci-fi"]
          }
       ]
    }
]
```

The resulting document in Meilisearch will be

```json
{
       "id": 1,
       "label": "UGC Paris 13",
       "movies": [
          {
              "title": "Dune",
              "genres": ["Sci-fi"]
          }
       ]
}
```

Reply:
Very helpful, thanks guys!